crmsh/corosync.py

```diff
@@ -811,20 +811,48 @@ class IPAlreadyConfiguredError(Exception):
     pass
 
 
-def add_node_ucast(IParray, node_id=None):
-
-    f = open(conf()).read()
-    p = Parser(f)
-
-    # to check if the same IP already configured
+def find_configured_ip(IParray):
+    """
+    find if the same IP already configured
+    If so, raise IPAlreadyConfiguredError
+    """
+    with open(conf()) as f:
+        p = Parser(f.read())
+
+    # get exist ip list from corosync.conf
     exist_iplist = []
-    for path in p.all_paths():
+    for path in set(p.all_paths()):
         if re.search('nodelist.node.ring[0-9]*_addr', path):
             exist_iplist.extend(p.get_all(path))
+
+    # all_possible_ip is a ip list to check whether one of them already configured
+    all_possible_ip = []
+    # get local ip list
+    local_ip_list = utils.ip_in_local(utils.is_ipv6(IParray[0]))
+    # extend all_possible_ip if IParray contain local ip
+    # to avoid this scenarios in join node:
+    #   eth0's ip already configured in corosync.conf
+    #   eth1's ip also want to add in nodelist
+    # if this scenarios happened, raise IPAlreadyConfiguredError
     for ip in IParray:
+        if ip in local_ip_list:
```

Comment:
I don't understand much this piece of code.
Here you check if some local IP is in the `IParray` list and if just 1 is there, you add the whole list.
Why do you add the whole list? Shouldn't you add just the found IP address?

By the way, you can replace this for and if by:
```
if bool(IParray & local_ip_list):
```

Reply:
Consider this scenario:
Join node eth0's ip `10.10.10.2` already configured in corosync.conf;
then join node eth1's ip `20.20.20.2` want to add in corosync.conf;
Local ip list in join node is `["10.10.10.2", "20.20.20.2"]`,
if we don't extend local ip list to `all_possible_ip`, no eception `IPAlreadyConfiguredError` will raise, `20.20.20.2` will add to corosync.conf

Reply:
But at the end you are adding all the IP addresses in `ip_list` to `all_possible_ip`, even though only one might be in both lists. For example:
```
ip_list = [1.1.1.1, 2.2.2.2]
local_ip_list = [1.1.1.1, 3.3.3.3, 4.4.4.4]
all_possible_ip = [1.1.1.1, 2.2.2.2, 1.1.1.1, 3.3.3.3, 4.4.4.4]
```
Only `1.1.1.1` matches but you add all of the addresses. Is this what you want? Just to understand better.

If this is what you want, and to make the code easier. I would just to this:
```
all_ip_list = set(ip_list)
if (set(ip_list) & set(local_ip_list)):
    all_ip_list |= set(local_ip_list)
if bool(all_possible_ip & set(corosync_ip_list)):
    raise IPAlreadyConfiguredError("IP {} was already configured".format(ip))
```
This piece of code does:
1. If some value in ip_list is in local_ip list -> all_ip_list = ip_list + local_ip_list, else all_ip_list = ip_list
2. If some value in all_ip_list is in corosync_ip_list, raise the error

Reply:
Changed.
Thank you @arbulu89

Reply:
Does it mean we don't mind if any IPs from ip_list are not in local_ip_list? We'd just try whatever the user is asking for? Not really a problem since there's fallback.  Just wondering :-)